### PR TITLE
pkg/k8s: use a single HTTP Client for all kube-apiserver connections

### PR DESCRIPTION
**Pull request by aanm**
The current heartbeat behavior is checking if one of the seven
connections, one for each resource type, is health. This prevents Cilium
from detecting that any of the remaining six connections is faulty which
might cause out-of-sync issues for some resources but not all. To
prevent this issue we can use a single HTTP Client for all resources,
including the heartbeat, which will make the detection of an unhealthy
connection more reliable.

Also, besides the bug-fix this commit is introducing, it is also
reducing the amount of connections open to kube-apiserver by 6 which
will reduce the amount of resources used by kube-apiserver.

Signed-off-by: André Martins <andre@cilium.io>

```release-note
Improve reliably of faulty connections for kube-apiservers behind a LB.
Reduce the number of connections to kube-apiserver by 6 for each cilium-agent.
```

### CODE REVIEW

File: pkg/k8s/client.go [hunk 1]
```diff
@@ -121,51 +121,40 @@ func createClient(config *rest.Config, cs kubernetes.Interface) error {
...
+		return fmt.Errorf("unable to create k8s client: %s", err)
 	}
 
 	k8sCiliumCLI.Interface = createdCiliumK8sClient
 
-	return closeAllConns, nil
+	return nil
 }
 
 func createAPIExtensionsClient() error {
```
Comment (sayboras) [inline pkg/k8s/client.go]: any reason not reusing http client for extensions client ? or is there no significant benefit?

**Review by sayboras — APPROVED** [review]
I have one one question as per below, but not a blocker.

LGTM :100:. Out of curiosity, any reason you are checking this part ? or just continuous improvement?

File: pkg/k8s/client.go (hunk 1 above)
Comment (aanm) [reply pkg/k8s/client.go]: oh this was something that I missed. Thanks for spotting it :male_detective:

**Review by sayboras — APPROVED** [review]

**Comment by aanm** [comment]
/test-1.23-net-next

File: pkg/k8s/init.go [hunk 2]
```diff
@@ -134,17 +135,32 @@ func useNodeCIDR(n *nodeTypes.Node) {
...
+	if err != nil {
+		return fmt.Errorf("unable to create k8s client rest configuration: %s", err)
+	}
+	closeAllDefaultClientConns := setDialer(restConfig)
+	// Use the same http client for all k8s connections. It does not matter that
+	// we are using a restConfig for the HTTP client that differs from each
+	// individual client since the rest.HTTPClientFor only does not use fields
+	// that are specific for each client, for example:
+	// restConfig.ContentConfig.ContentType.
+	httpClient, err := rest.HTTPClientFor(restConfig)
```
Comment (ldelossa) [inline pkg/k8s/init.go:148]: Should we stash this httpClient somewhere global? 

If the need arises to instantiate another K8s resource, like a client, somewhere else in code, having this shared client available will help us not spam more clients across the code base. 

This is a moot point if we only expect decoupled code to just go and grab the Cilium and K8s clients from our k8s package and not create any kind of K8s clients themselves.

**Review by ldelossa — APPROVED** [review]
💯 refactor

File: pkg/k8s/init.go (hunk 2 above)
Comment (christarazi) [reply pkg/k8s/init.go:148]: > This is a moot point if we only expect decoupled code to just go and grab the Cilium and K8s clients from our k8s package and not create any kind of K8s clients themselves.

I think the expectation is that any Cilium code interested in interacting with K8s will import `pkg/k8s` to use the "blessed" clients. Anything else should be avoided.

**Review by christarazi — APPROVED** [review]

**Comment by aanm** [comment]
The current failures are already know and being addressed. Merging.

**Pull request merged**